tools/parallelize/parallelize.cpp

```diff
@@ -0,0 +1,203 @@
+            parent->update_display();
+        }
+
+        void schedule(std::size_t command) {
+            commandRunning = command;
+            executive.begin_execution(nullptr, parent->commands[command].first.data(), 0, nullptr);
+            tpWait.wait_for(executive.get_wait_handle());
+        }
+    };
+
+    void update_display() noexcept {
+        const auto commandsSize = commands.size();
+        const auto next         = nextCommandToRun;
+        const auto running      = runningConcurrency.load(std::memory_order_relaxed);
+        assert(running <= next);
+        printf("%zu scheduled %zu completed %zu running\n", commandsSize, next - running, running);
+    }
+};
+
+inline void schedule_command(parallelizer& p, const std::wstring_view commandPrefix, const std::wstring& native) {
+    std::wstring toExecute;
+    toExecute.reserve(commandPrefix.size() + native.size() + 1);
+    toExecute.assign(commandPrefix);
+    toExecute.push_back(L' ');
+    toExecute.append(native);
+    p.add_command(std::move(toExecute));
+}
+
+extern "C" int wmain(int argc, wchar_t* argv[]) {
+    try {
+        if (argc < 3) {
+            puts("Usage: parallelize.exe commandPrefix pathRoots");
+            return 1;
+        }
+
+        parallelizer p;
+        std::wstring_view commandPrefix(argv[1]);
+        for (int idx = 2; idx < argc; ++idx) {
+            std::filesystem::path thisSpec(argv[idx]);
+            if (std::filesystem::is_regular_file(thisSpec)) {
+                schedule_command(p, commandPrefix, thisSpec.native());
+                continue;
+            }
+
+            for (const auto& dirEntry : std::filesystem::recursive_directory_iterator(std::move(thisSpec))) {
+                if (!dirEntry.is_regular_file()) {
+                    printf("Skipping non-regular-file %ls\n", dirEntry.path().c_str());
+                    continue;
+                }
+
+                schedule_command(p, commandPrefix, dirEntry.path().native());
+            }
+        }
+
+        p.wait_all();
+        auto& ran          = p.results();
+        const auto success = [](const std::pair<std::wstring, std::optional<execution_result>>& command) noexcept {
+            auto& result = command.second.value();
+            return result.exitCode == 0 && result.output.empty();
+        };
+
+        const std::size_t successes = std::count_if(ran.begin(), ran.end(), success);
+        printf("%zu commands ran successfully.\n", successes);
+        if (successes == ran.size()) {
+            return 0;
+        }
+
+        for (auto& command : ran) {
+            if (success(command)) {
+                continue;
+            }
+
+            auto& result = command.second.value();
+            printf("%ls exited with 0x%08lX and output:\n%s\n", command.first.c_str(), result.exitCode,
+                result.output.c_str());
+        }
+
+        return 1;
+    } catch (std::filesystem::filesystem_error& err) {
+        fputs(err.what(), stderr);

```

Comment:
It's somewhat confusing that there are multiple ways for this to fail, but only some of them print to stderr.

Reply:
They both print to stderr (and call abort()), I'm not sure I understand your comment.

Reply:
I mean, if a test fails, you use `printf()` to `stdout`. But if the directory traversal fails, you print to `stderr`. I suppose one is "worse", but it seems inconsistent.

Reply:
Right, one of those is a failure of `parallelize.exe`, and one of them is normal operation. (For example our tests return 100 on "success", so that's not a failure)

All things that are actually failures in this program goto abort.